Intuos paper support v1

whot opened the pull request:
Definitely wip-do not merge, this is just so it's backed up somewhere.

Biggest change here is that we can't use the protocol subclassing as-is, we need an "unknown protocol" object first that we can use to issue the first connection attempt, based on the reply we know what subclass we actually have.

- [x] registration of device
- [x] reading the device name - *for some reason this is a 0 string*
- [x] connection of device after registration
- [x] download stroke data
- [x] interpret stroke data

whot committed:
wacom: add support for the Intuos Pro 2 paper

bentiss commented on tuhi/wacom.py:
```diff
@@ -99,6 +99,11 @@ class WacomProtocol(GObject.Object):
+    communication patterns. Any device-specific quirks are handled in the
+    subclasses of this device, reverse-ordered by date. Right now, this
+    means WacomProtocol represents the IntuosPro, subclassing that is the
+    Slate which itself is subclassed by the Spark.
```
After a little bit of thinking, I think we should do this the other way around.
Having the older first and the newer inherits from their parents.

The reason is that with a reverse ordering, we will have to check on every generation when adding a new device. While if we have it the older first, adding a new device protocol will not require us to change anything in the older devices, and so we will not have to test those but the new ones.

bentiss commented:
FYI, I just pushed a rebase of this branch on to https://github.com/bentiss/tuhi/tree/wip/intuos-paper

There are minor differences from yours like error messages.

Now, I am starting to rework on the last few patches to have something saner.

bentiss commented:
See https://github.com/bentiss/tuhi/tree/wip/intuos-paper-v2 for an update of this PR based on #78

bentiss commented on tuhi/wacom.py:
```diff
@@ -608,7 +608,7 @@ class WacomProtocolSlate(WacomProtocolSpark):
 
     def __init__(self, device, uuid):
-        super().__init__(device)
+        super().__init__(device, uuid)
```
Ouch, how could I have missed that one :/

bentiss commented on tuhi/wacom.py:
```diff
@@ -316,18 +316,27 @@ def e3_command(self):
                                       expected_opcode=0xb3)
 
-    def set_time(self):
+    def time_to_bytes(self):
```
I like this one, so I cherry-picked it on top of master and stripped out the Intuos Pro changes that can now be squashed in the initial support for the Intuos Pro commit

bentiss commented on tuhi/wacom.py:
```diff
@@ -316,18 +316,27 @@ def e3_command(self):
+        return [int(i) for i in binascii.unhexlify(current_time)]
+
+    def time_from_bytes(self, data):
+        assert len(data >= 6)
```
typo: `len(data) >= 6` (fixed locally)

bentiss commented on tuhi/ble.py:
```diff
@@ -145,10 +145,11 @@ def connected(self):
         md = self.interface.get_cached_property('ManufacturerData')
-        if md is not None and vendor_id in md.keys():
-            return md[vendor_id]
+        if md is not None and self.vendor_id in md.keys():
```
I am not a big fan of this one:
self.vendor_id already gets the cached property `'ManufacturerData'`, and so if `vendor_id` is not `None`, we already know that it will be in `md.keys()`.

I'll try to have a more straightforward commit.

bentiss replied:
OK, pushed a pythonic version of this on master:
   6cff3b8..46ec85d  master -> master

bentiss reviewed (commented):
I have cherry-picked the cleanup commits and pushed them to master in addition to the split time parsing function one that is amended.
   c2bba6b..6cff3b8  master -> master

whot committed:
wacom: merge two debug strings

whot committed:
wacom: factor out the prefix handling

The Intuos Pro has a different prefix, we'll need to override this. Dropping
the comment about the 8bt, on the intuos pro the prefix is nonsensical so
let's not leave a false lead there.

whot committed:
wacom: skip over the prefix when handling pen data

The prefix (4 bytes on the spark/slate) is some data that we don't know yet.
It *could* be the size of a tablet/drawing in LE byte order: 14434x29794
although that doesn't quite match the Intuos Pro data where the prefix is
different.

Either way, having this be an opcode of 0x3800 was probably just coincidence.
Let's skip over the header normally and assume that once we have a prefix, we
have a drawing. This opcode never occurs a second time within the same
drawing.

whot committed:
wacom: factor out the 'next stroke data' parsing

whot committed:
wacom: fix Spark CRC check

All it needs is less reversing

whot committed:
wacom: append multiple nordic answers to each other

On the Spark, the 0xc9 CRC comes almost immediately after the 0xc8
end-of-sequence. This causes a race condition where sometimes we process the
0xc8 before the 0xc9 arrives, other times the 0xc9 overwrites the current
nordic_answer value before we get to it - triggering an unexpected opcode
exception.

Fix this by appending the nordic answer and only ever pulling off enough to
satisfy the current request.

Fixes #90

whot committed:
wacom: make relative/abs events more obvious in the log

Mark the deltas with a * if they're the ones that changed.

whot committed:
wacom: don't add points to the log until we have an absolute value

On the Intuos Pro at least we sometimes get events that are all relative at
first and it can take several events for us to have all three values as
absolute. This is likely a parsing error on our side, but meanwhile don't
write the data until we have at least one known value for all three axes.

And because the SVG writer isn't happy with empty strokes, add a seal()
function to the drawing to purge empty strokes.

Fixed #92 or at least works around it

whot committed:
wacom: we only have one drawing per exchange

skip putting a single drawing in a list

whot commented:
Ready to go!

Only the last patch is intuos support, the rest is cleanup and shuffling around things again

bentiss commented:
\o/
